File: leetcode/3.无重复字符的最长子串.java
/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0, r = -1;
        int[] freq = new int[256];
        int res = 0;
        while (l<s.length()) {
            if ( r+1 < s.length()&&freq[s.charAt(r+1)]==0) {
                freq[s.charAt(++r)]++;
            }else{
                freq[s.charAt(l++)]--;
            }
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}
// @lc code=end

File: leetcode/3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	str := []byte(s)
	record := make(map[byte]int)
	start := 0
	maxLength := 0
	for i, ch := range str {
		if lastI, ok := record[ch]; ok && lastI >= start {
			start = lastI + 1
		}
		if i-start+1 > maxLength {
			maxLength = i - start + 1
		}
		record[ch] = i
	}
	return maxLength
}

// @lc code=end

